packages/components/src/app/word-query/word-query-adapter.js

```diff
@@ -0,0 +1,115 @@
+import { ApolloClient, InMemoryCache, HttpLink, gql } from '@apollo/client/core'
+
+let singleInstance
+
+const WORD_QUERY =
+  gql`
+    query WordQuery($langCode: Language, $userID: String) {
+      word(langCode: $langCode, userID: $userID) @client
+    }
+  `
+
+/**
+ * Converts a results of GraphQL word query into JS objects.
+ * Provides several methods to retrieve data via GraphQL word query.
+ */
+export default class WordQueryAdapter {
+  constructor ({ wordQueryResolver } = {}) {
+    const resolvers = {}
+
+    const cache = new InMemoryCache({
+      typePolicies: {
+        Query: {
+          fields: {
+            word: {
+              read: wordQueryResolver
+            }
+          }
+        }
+      }
+    })
+
+    this._gql = new ApolloClient({
+      cache,
+      link: new HttpLink({ uri: '/graphql', fetch }),
+      resolvers
+    })
+  }
+
+  static getInstance () {
+    if (!singleInstance) {
+      singleInstance = new WordQueryAdapter()
+    }
+    return singleInstance
+  }
+
+  /**
+   * An asynchronous GraphQL query. Data is returned when the query is fully completed.
+   *
+   * @param {object} options - Data that will be passed to the GraphQL query.
+   * @param {object} options.variables - Variables to be passed to the GraphQL query.
+   * @returns {Promise<*>} - A promise that is resolved with the GraphQL result.
+   */
+  async query ({ variables } = {}) {
+    const gqlQuery = {
+      query: WORD_QUERY,
+      variables
+    }
+    return this._gql.query(gqlQuery)
+  }
+
+  /**
+   * An asynchronous GraphQL query. Data is returned when the query is fully completed.
+   *
+   * @param {object} options - Data that will be passed to the GraphQL query.
+   * @param {object} options.variables - Variables to be passed to the GraphQL query.
+   * @param {Function} options.dataCallback - A function to be called every time new data arrives.
+   * @param {number} options.pollInterval - An interval in milliseconds specifying how often a new data will be retrieved.
+   * @param {number} options.maxIterations - How many iterations to do before stopping the query.
+   */
+  observableQuery ({ variables, dataCallback, pollInterval = 1000, maxIterations = 10 } = {}) {
```

Comment:
What I've meant was that it is asynchronous by nature but it does not fit the JS `async` specification. The `observableQuery` is created for situations when not all query results became available immediately but may be delivered over time. In our case, for example, it can take a lot of time to retrieve all lexical data, but some results (like lexemes) may be delivered to the client before the query is fully complete (i.e. until short definitions are obtained).

The `observableQuery` is aware of the dynamic nature of the query and it calls the user-provided data callback function (`dataCallback`) every time when query results are updated. In other words, `observableQuery` knows that the query is "live" and it polls the query data every X milliseconds or until the maximum number of poll attempts is reached (those numbers needs careful tweaking for production).

Do you think there could be some other way to describe it in order to avoid confusion with the JS `async` specs?

Reply:
I think this says it very well :-) " observableQuery knows that the query is "live" and it polls the query data every X milliseconds or until the maximum number of poll attempts is reached"

Reply:
Great, will put that into description.